src/mlmc/quantity_concept.py

```diff
@@ -121,120 +325,451 @@ def mult_op(x, y):
+
+        q_type = copy.deepcopy(self._qtype)
+        position = self._times.index(key)
+        q_type.start = position * q_type.size()
+        return q_type
+
+
+class FieldType(QType):
+    def __init__(self, args: List[Tuple[str, QType]], start=0):
+        """
+        QType must have same structure
+        :param args:
+        """
+        self._dict = dict(args)
+        self._qtype = args[0][1]
+        self.start = start
+        assert all(q_type == self._qtype for _, q_type in args)
+
+    def size(self) -> int:
+        return len(self._dict.keys()) * self._qtype.size()
+
+    def __getitem__(self, key):
+        if key not in self._dict:
+            raise KeyError("Key " + str(key) + " was not found in FieldType" +
+                           ". Available keys: " + str(list(self._dict.keys())))
+
+        q_type = copy.deepcopy(self._qtype)
+        position = list(self._dict.keys()).index(key)
+        q_type.start = position * q_type.size()
+        return q_type
+
+    def __copy__(self):
+        new = type(self)([(k, v) for k, v in self._dict.items()])
+        new.__dict__.update(self.__dict__)
+        return new
+
+
+class DictType(QType):
+    def __init__(self, args: List[Tuple[str, QType]]):
+        self._dict = dict(args)
+        self.start = 0
+
+        self._check_base_type()
+
+    def _check_base_type(self):
+        qtypes = list(self._dict.values())
+        assert all(isinstance(qtype.base_qtype(), type(qtypes[0].base_qtype())) for qtype in qtypes[1:]), \
+            "All QTypes must have same base QType, either SacalarType or BoolType"
+
+    def base_qtype(self):
+        return list(self._dict.values())[0].base_qtype()
+
+    def size(self) -> int:
+        return int(np.sum(q_type.size() for _, q_type in self._dict.items()))
+
+    def get_qtypes(self):
+        return self._dict.values()
+
+    def replace_scalar(self, new_qtype):
+        for key, qtype in self._dict.items():
+            if isinstance(qtype, ScalarType):
+                self._dict[key] = new_qtype
+            else:
+                qtype.replace_scalar(new_qtype)
 
+    def __getitem__(self, key):
+        if key not in self._dict:
+            raise KeyError("Key " + str(key) + " was not found in DictType" +
+                           ". Available keys: " + str(list(self._dict.keys())))
 
+        q_type = self._dict[key]
 
-class Array(QBase):
-    def __init__(self, *args):
-        assert( all( (a.is_array() for a in args) ) )
-        assert( all( (a.shape == args[0].shape for a in args) ) )
-        self.q_list = args
-        self.shape = [len(self.q_list)] + args[0].shape
+        size = 0
+        for k, qt in self._dict.items():
+            if k == key:
+                break
+            size += qt.size()
 
+        q_type.start = size
 
+        return q_type
 
-class Field(QBase):
-    def __init__(self, *args: Tuple[str, QBase]):
-        self.quantity_dict = OrderedDict(*args)
 
+def custom_copy(quantities):
```

Comment:
Possibly make it a private static method of Quantity and use appropriate name. 'custom' says nothing.